jax/interpreters/xla.py

```diff
@@ -149,23 +149,39 @@ def _make_abstract_python_scalar(typ, _):
 
 def apply_primitive(prim, *args, **params):
   """Impl rule that compiles and runs a single primitive 'prim' using XLA."""
-  abstract_args = map(abstractify, args)
-  compiled_fun = xla_primitive_callable(prim, *abstract_args, **params)
+  compiled_fun = xla_primitive_callable(prim, *map(arg_spec, args), **params)
   return compiled_fun(*args)
 
+def arg_spec(x):
+  aval = abstractify(x)
+  try:
+    return aval, x._device
+  except:
+    return aval, None
+
 @cache()
-def xla_primitive_callable(prim, *abstract_args, **params):
+def xla_primitive_callable(prim, *arg_specs, **params):
```

Comment:
Doesn't `Device` already need to be hashable for the compilation cache to work here?

Reply:
The `Device` class needs to be hashable (on id) but not instances of it.

Reply:
Do we have a test that makes sure primitive compilation stays cached?

Reply:
Yes I believe so. But in any case the type itself isn't being redefined.

Reply:
Just to be sure, I added a new one.